nixos/nginx: don't hide nginx config errors on nixos-rebuild --switch with reload enabled

danbst opened the pull request:
Closes https://github.com/NixOS/nixpkgs/issues/73455

###### Things done


- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @domenkozar @Izorkin @aanderse @ckauhaus

danbst committed:
nixos/nginx: don't hide nginx config errors on nixos-rebuild --switch
with reload enabled

Closes https://github.com/NixOS/nixpkgs/issues/73455

danbst committed:
nixos/nginx: don't hide nginx config errors on nixos-rebuild --switch
with reload enabled

Closes https://github.com/NixOS/nixpkgs/issues/73455

danbst commented:
I know about https://github.com/NixOS/nixpkgs/pull/74072, but I have a test working so it is not blocked by that PR.

domenkozar reviewed (commented):
:heart_eyes:

Mic92 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -705,11 +707,15 @@ in
       script = ''
         if ${pkgs.systemd}/bin/systemctl -q is-active nginx.service ; then
-          ${pkgs.systemd}/bin/systemctl reload nginx.service
+          ${execCommand} -s reload && \
```
Is the reload command following this still necessary if this command already performs a reload?

danbst committed:
few fixes

danbst committed:
few fixes

danbst committed:
another fix for extra delay during switch

danbst committed:
another fix for extra delay during switch

danbst commented:
@Mic92 you are right, the `systemctl reload` was redundant. However I now changed it a bit - it doesn't reload nginx, but only checks config file for errors.

The reason I still do `systemctl reload` is to create an entry in log file for `nginx.service`:
```
# journalctl -u nginx | tail -n2
Dec 23 09:35:07 c1 systemd[1]: Reloading Nginx Web Server.
Dec 23 09:35:07 c1 systemd[1]: Reloaded Nginx Web Server.
```
I can implement that with running logger manually from reload script. Do you think it is better to run logger explicitly, rather than relying on `systemd reload`?

I also removed `stopIfChanged = false;`. Due to how nixos activation works, on nginx confiuration change it tries to a) restart `nginx-config-reload.service` and b) start `multi-user.target`. If there is an error in config, it will fail twice - first time during restart, second time during start, because those are done as separate systemctl commands. In some cases (DNS lookup error) this may cause very long activation delay.

However, if unit is stopped first, then activate starts both `nginx-config-reload.service` and `multi-user.target` together, which produces only one failure.

ckauhaus commented:
@danbst Looks good to me. Please rebase to current master.

danbst committed:
Merge branch 'master' into nginx-reload-warn-errors

danbst committed:
Merge branch 'nginx-reload-warn-errors' of github.com:danbst/nixpkgs into nginx-reload-warn-errors

danbst commented:
@domenkozar @ckauhaus what do you think, this has to be backported to 19.09? I believe it should, but I need confirmation

domenkozar commented:
Yup.

Mic92 commented on nixos/modules/services/web-servers/nginx/default.nix:
```diff
@@ -706,11 +708,18 @@ in
       script = ''
         if ${pkgs.systemd}/bin/systemctl -q is-active nginx.service ; then
-          ${pkgs.systemd}/bin/systemctl reload nginx.service
+          ${execCommand} -t && \
+            ${pkgs.systemd}/bin/systemctl reload nginx.service
         fi
       '';
```
This line breaks nginx. It causes directories in `/var/cache/nginx/` to be owned by `nobody`, which breaks large POST uploads. I think this might due the fact that nginx is no longer started as root. cc @Izorkin

Izorkin replied:
#85820 - fix variant and https://github.com/NixOS/nixpkgs/pull/85820#issuecomment-620378045

Mic92 replied:
Ok. I have a simpler version. Will post.

Mic92 replied:
https://github.com/NixOS/nixpkgs/pull/95249
